proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,385 @@
+    
+  var wrappedValue: Value {
+    get 
+    nonmutating set
+  }
+    
+  var projectedValue: Self {
+    self
+  }
+}
+
+typealias A = (Reference<Int>) -> Void
+
+let a: A = { @Reference foo in
+  ...
+}
+```
+
+    Becomes:
+
+```swift
+let a: A = { _foo in
+  var foo: Int {
+    get { 
+      _foo.wrappedValue 
+    }
+    set { 
+      _foo.wrappedValue = newValue
+    }
+  }
+
+  var $foo: Int {
+    get { 
+      _foo.projectedValue 
+    }
+  }
+    
+  ...
+}
+```
+
+2. Value Semantics Wrapper
+
+```swift
+@propertyWrapper
+struct Wrapper<Value> {
+  init(wrappedValue: Value) 
+    
+  var wrappedValue: Value {
+    get 
+    set
+  }
+}
+
+typealias B = (inout Wrapper<Int>) -> Void
+
+let b: B = { @Wrapper foo in
+  ...
+}
+```
+
+    Becomes:
+
+```swift
+let b: B = { _foo in
+  var foo: Int {
+    get { 
+      _foo.wrappedValue 
+    }
+    set { 
+      _foo.wrappedValue = newValue
+    }
+    // Since the paramter is marked `inout`
+    // we are allowed to have a mutating setter.
+  }
+  
+  ...
+}
+```
+
+3. Value Semantics Wrapper with Special Initializer
+
+```swift
+@propertyWrapper
+class WrapperObject<Value> {
+  init(wrappedValue: Value) 
+    
+  var wrappedValue: Value {
+    get 
+    set 
+    // Not actually mutating
+    // because `WrapperObject` is
+    // a class.
+  }
+}
+
+typealias C = (WrapperObject<Int>) -> Void
+
+let c: C = { @WrapperObject foo in
+  ...
+}
+```
+
+    Becomes:
+
+```swift
+let c: C = { _foo in
+  var foo: Int {
+    get { 
+      _foo.wrappedValue 
+    }
+    set { 
+      _foo.wrappedValue = newValue
+    }
+    // Since `WrapperObject` has reference
+    // semantics we can include a setter.
+  }
+
+  ...
+}
+```
+
+
+## Source compatibility
+
+This is an additive change with _no_ impact on **source compatibility**.
+
+
+## Effect on ABI stability
+
+This is an additive change with _no_ impact on **ABI stability**.
+
+
+## Effect on API resilience
+
+This is an additive change with _no_ impact on **API resilience**.
+
+
+## Alternatives Considered
+
+### Infer Property Wrappers in Closure Parameters 
+
+TBD
+
+
+### Support `@autoclosure` and `@escaping` in Function Parameters
+
+TBD
+
+
+## Future Directions
+
+### Support property wrapper initialization from a projected value
+
+Today, a property wrapper can be initialized from an instance of its `wrappedValue` type if the wrapper provides a suitable `init(wrappedValue:)`. The same initialization strategy is used in this proposal for property wrapper parameters to allow users to pass a wrapped value as a property wrapper argument. We could extend this model to support initializing a property wrapper from an instance of its `projectedValue` type by allowing property wrappers to define an `init(projectedValue:)` that follows the same rules as `init(wrappedValue:)`. This could allow users to additionally pass a projected value as a property wrapper argument, like so:
+
+```swift
+@propertyWrapper
+struct Wrapper<T> {
```

Comment:
I think we should rename the generic parameter `T` to `Value` for better clarity.